### DISCUSSION TITLE
How do I write a syntax injection?

### DISCUSSION THREAD

**Discussion by NullVoxPopuli** [nvim-treesitter/nvim-treesitter#1171] [Q&A] [answered]
Hello! 👋 

I'm wanting to add a nested syntax to javascript / typescript, [shown here](https://github.com/ember-template-imports/ember-template-imports#using-template-tags-and-gjs-files)
example:
```jsx
// components/hello.gjs
<template>
  <span>Hello, {{@name}}!</span>
</template>
```
which, is JSX-esque, sortof. the `<template>` tag doesn't go in a function though, and everything within `<template>` is Glimmer syntax

I tried this:
```diff
❯ git diff --staged
diff --git a/queries/jsx/injections.scm b/queries/jsx/injections.scm
new file mode 100644
index 0000000..a24f1b6
--- /dev/null
+++ b/queries/jsx/injections.scm
@@ -0,0 +1,2 @@
+((jsx_opening_element
+  (#eq? @_name "template") @glimmer))
```
_but_, I don't think this is correct, because when symlink the jsx to neovim's queries directory, I get an error. Also, I don't know what tags / variables are available... is there a coc plugin for scheme / nvim-treesitter?
This is the error I get:
![image](https://user-images.githubusercontent.com/199018/114250545-0d161200-996c-11eb-9754-27f3b98da984.png)

So, yeah, mostly what I'm after:
 - how do I know what I can do?
 - how do I get more meaningful errors? (to help self-discovery?)
 
 I've read: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#injections, but I didn't feel it that helpful, or maybe I don't have the base knowledge for it to make sense? idk.

**Comment by theHamsta** [comment, marked as answer]
```
(jsx_element
  (jsx_opening_element) @_name (#eq? @_name "<template>")) @glimmer
```

**Reply by theHamsta** [reply]
Your query is very short but has multiple issues

- the text you want to match is `<template>`
- `(#eq? @_name "template")` is not a node so you can't put behind it `@glimmer`
- `(#eq? @_name "template")` you can't use `@_name` when you didn't associate it with a node

**Reply by NullVoxPopuli** [reply]
thanks!

I've learned that I probably can't extend TSX, but I need to add a new parser for this! progress!

![image](https://user-images.githubusercontent.com/199018/114253106-ae09ca80-9976-11eb-9187-931c3a82b806.png)

**Reply by NullVoxPopuli** [reply]
and/_or_ the language didn't switch 🤔

**Reply by NullVoxPopuli** [reply]
did I do something wrong? how come it thinks everything between the template tags is still tsx?

**Comment by stsewd** [comment]
> is there a coc plugin for scheme / nvim-treesitter?

The playground + query linter is useful when writing queries https://github.com/nvim-treesitter/playground#query-linter

**Reply by NullVoxPopuli** [reply]
🤔 is there a command to open the query linter? my `o` key opens a new buffer above :-\

**Reply by theHamsta** [reply]
The query linter will automatically be active in query buffers when you activated it. Either in you query buffer or in `javascript/injections.scm`. Unfortunatelly it will not work in `jsx/injections.scm` because there is no jsx parser.

**Reply by theHamsta** [reply]
The query that I posted works nicely on my machine. I tested it in `javascript/injections.scm` and copied it to `jsx/injections.scm` when it was ok.

**Reply by NullVoxPopuli** [reply]
🤔 there is a glimmer parser (in another repo), and I don't see query buffers either. So is that the issue? the parser has to be in the same repo as the queries?

**Reply by theHamsta** [reply]
It must guess somehow what is the parser for this queries. It knows that for the playground buffer that you open with `o` and when the folder is equal to the parser name.

**Reply by theHamsta** [reply]
We could extend the heurisitc to `tree-sitter-<parsername>/queries/foo.scm`

**Reply by theHamsta** [reply]
Or allow an annotation in the query file.

**Reply by theHamsta** [reply]
Or read the parser name from `grammar.js`

**Reply by NullVoxPopuli** [reply]
Hey, I'm now getting stuff like this!
![image](https://user-images.githubusercontent.com/199018/114253250-79e2d980-9977-11eb-9cf7-7c7cd5844fc4.png)
neat
